File: DSA implemented in Java/Anchal/Binary Trees/2) Check for Balanced Tree.py
"""
Check for Balanced Tree

Easy

Given a binary tree, find if it is height balanced or not. 
A tree is height balanced if difference between heights of left and right subtrees is not more than one for all nodes of tree. 

A height balanced tree
        1
     /     \
   10      39
  /
5

An unbalanced tree
        1
     /    
   10   
  /
5

Example 1:
Input:
      1
    /
   2
    \
     3 
Output: 0
Explanation: The max difference in height
of left subtree and right subtree is 2,
which is greater than 1. Hence unbalanced

Example 2:
Input:
       10
     /   \
    20   30 
  /   \
 40   60
Output: 1
Explanation: The max difference in height
of left subtree and right subtree is 1.
Hence balanced. 
Your Task:
You don't need to take input. Just complete the function isBalanced() that takes root node as parameter and returns true, if the tree is balanced else returns false.

Constraints:
1 <= Number of nodes <= 10^5
1 <= Data of a node <= 10^9

Expected time complexity: O(N)
Expected auxiliary space: O(h) , where h = height of tree

"""


# SOLUTION

#DFS

'''
class Node: 
    # Constructor to create a new Node 
    def __init__(self, data): 
        self.data = data 
        self.left = None
        self.right = None
'''


#Function to check whether a binary tree is balanced or not.
class Solution:
    def isBalanced(self, root):
        # Check if the height difference is not -1 (indicating imbalance)
        return self.dfsHeight(root) != -1

    def dfsHeight(self, root):
        # Base case: leaf node (height is 0)
        if root is None:
            return 0

        # Recursive calls to find the heights of left and right subtrees
        lh = self.dfsHeight(root.left)
        rh = self.dfsHeight(root.right)

        # If either subtree is unbalanced, propagate the unbalanced status
        if lh == -1 or rh == -1:
            return -1

        # Check if the current subtree is balanced
        if abs(rh - lh) > 1:
            return -1

        # Return the height of the current subtree
        return 1 + max(lh, rh)
        
'''
Time Complexity:
The time complexity of the code is O(N), where N is the number of nodes in the binary tree. This is because each node is visited once, and for each node, constant time operations are performed.


Space Complexity:
The space complexity is O(H), where H is the height of the binary tree. In the worst case, the recursion depth would be equal to the height of the tree. Therefore, the space complexity is determined by the maximum depth of the recursion stack.

In the worst case, if the binary tree is unbalanced (skewed), the height becomes N (number of nodes), and the space complexity becomes O(N). In the best case, if the tree is balanced, the height is log(N) for a binary tree with N nodes, resulting in a space complexity of O(log(N)).

So, the overall space complexity is O(H), where H is the height of the binary tree.
'''


# BFS

from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isBalanced(self, root):
        if not root:
            return True  # An empty tree is considered balanced

        queue = deque([(root, 1)])  # Each element in the queue is a tuple (node, depth)

        while queue:
            node, depth = queue.popleft()

            left_depth = self.get_depth(node.left)
            right_depth = self.get_depth(node.right)

            if abs(left_depth - right_depth) > 1:
                return False

            if node.left:
                queue.append((node.left, depth + 1))

            if node.right:
                queue.append((node.right, depth + 1))

        return True

    def get_depth(self, node):
        if not node:
            return 0
        return 1 + max(self.get_depth(node.left), self.get_depth(node.right))

File: DSA implemented in Java/Anchal/Binary Trees/2) Check for Balanced Tree.java
/*
Check for Balanced Tree

Easy

Given a binary tree, find if it is height balanced or not. 
A tree is height balanced if difference between heights of left and right subtrees is not more than one for all nodes of tree. 

A height balanced tree
        1
     /     \
   10      39
  /
5

An unbalanced tree
        1
     /    
   10   
  /
5

Example 1:
Input:
      1
    /
   2
    \
     3 
Output: 0
Explanation: The max difference in height
of left subtree and right subtree is 2,
which is greater than 1. Hence unbalanced

Example 2:
Input:
       10
     /   \
    20   30 
  /   \
 40   60
Output: 1
Explanation: The max difference in height
of left subtree and right subtree is 1.
Hence balanced. 
Your Task:
You don't need to take input. Just complete the function isBalanced() that takes root node as parameter and returns true, if the tree is balanced else returns false.

Constraints:
1 <= Number of nodes <= 10^5
1 <= Data of a node <= 10^9

Expected time complexity: O(N)
Expected auxiliary space: O(h) , where h = height of tree

*/


// SOLUTION
/* 
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}

// Depth-First Search (DFS)

class Solution {
    public boolean isBalanced(TreeNode root) {
        return dfsHeight(root) != -1;
    }

    private int dfsHeight(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int lh = dfsHeight(root.left);
        if (lh == -1) {
            return -1;
        }

        int rh = dfsHeight(root.right);
        if (rh == -1) {
            return -1;
        }

        if (Math.abs(lh - rh) > 1) {
            return -1;
        }

        return 1 + Math.max(lh, rh);
    }
}
*/

// Breadth-First Search (BFS)

class Solution {
    public boolean isBalanced(TreeNode root) {
        if (root == null) {
            return true;  // An empty tree is considered balanced
        }

        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);

        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            int leftDepth = getDepth(node.left);
            int rightDepth = getDepth(node.right);

            if (Math.abs(leftDepth - rightDepth) > 1) {
                return false;
            }

            if (node.left != null) {
                queue.add(node.left);
            }

            if (node.right != null) {
                queue.add(node.right);
            }
        }

        return true;
    }

    private int getDepth(TreeNode node) {
        if (node == null) {
            return 0;
        }
        return 1 + Math.max(getDepth(node.left), getDepth(node.right));
    }
}


// The TreeNode class in Java represents a node in the binary tree, similar to the TreeNode class in Python.

// DFS Approach:

// The Solution class contains the isBalanced method and the dfsHeight helper method.
// isBalanced checks if the height difference is not -1, indicating an imbalance.
// dfsHeight recursively calculates the height of the left and right subtrees. If an imbalance is detected, it returns -1.

// BFS Approach:

// The Solution class also provides a BFS approach with the isBalanced and getDepth methods.
// isBalanced uses a queue to traverse the tree level by level.
// getDepth is a helper method that computes the depth of a given node.